### DISCUSSION TITLE
Implementing real-time video recommendations with websockets and Tensorflow lite in Serverpod

### DISCUSSION THREAD

**Discussion by snfxgsx** [serverpod/serverpod#2507] [Q&A] [answered]
Hi Serverpod Team,

I'm working on a short video app and am considering using WebSockets to provide real-time video recommendations to enhance user engagement. My plan is to handle user interactions (like likes, shares, and skips) on the client side and use WebSockets to send these interactions to the server. The server would then process these interactions and send back real-time video recommendations to the client.

Additionally, I plan to use TensorFlow Lite on the client side for the recommendation logic. This way, I can leverage machine learning for personalized recommendations while keeping the latency low by processing data on the client device.

On the server side, I'm thinking of setting up a WebSocket endpoint in Serverpod to handle incoming connections and messages from clients. This endpoint would process user interactions and potentially send back aggregated data or triggers for model updates. On the client side, I'll use the `web_socket_channel` package in a Flutter app to maintain an open WebSocket connection with the server. The client would send user interaction data to the server via WebSocket and listen for incoming messages from the server, which may include updates or new recommendation triggers.

For generating recommendations locally, I'll use TensorFlow Lite models. The app will capture user interactions and use them to generate recommendations locally, while also sending interaction data to the server for logging and additional processing if necessary.

Here is a simplified example of what I have in mind:

Server-Side WebSocket Endpoint:

import 'package:serverpod/serverpod.dart';

class VideoRecommendationEndpoint extends Endpoint {
  @override
  Future<void> initialize() async {
    await super.initialize();
  }

  @override
  void onConnect(Session session) {
    print('Client connected: ${session.id}');
  }

  @override
  void onDisconnect(Session session) {
    print('Client disconnected: ${session.id}');
  }

  @override
  void onMessage(Session session, dynamic message) {
    print('Received message from ${session.id}: $message');
  }
}

Client-Side WebSocket and TensorFlow Lite Integration:

import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:tflite_flutter/tflite_flutter.dart';

class WebSocketService {
  final WebSocketChannel _channel;

  WebSocketService(String url) : _channel = WebSocketChannel.connect(Uri.parse(url));

  void sendMessage(dynamic message) {
    _channel.sink.add(message);
  }

  Stream<dynamic> get messages => _channel.stream;

  void close() {
    _channel.sink.close();
  }
}

class InteractionHandler {
  final WebSocketService _webSocketService;
  final Interpreter _interpreter;

  InteractionHandler(this._webSocketService, this._interpreter);

  void handleInteraction(UserInteraction interaction) {
    var message = {
      'userId': 'user-id',
      'interactionType': interaction.runtimeType.toString(),
      'timestamp': DateTime.now().toIso8601String(),
      'data': interaction.data,
    };

    var recommendation = generateRecommendation(interaction);
    
    _webSocketService.sendMessage(message);
  }

  void listenForMessages() {
    _webSocketService.messages.listen((message) {});
  }

  dynamic generateRecommendation(UserInteraction interaction) {
    var input = [interaction.data];
    var output = List.filled(1, 0).reshape([1, 1]);
    _interpreter.run(input, output);
    return output[0];
  }
}

Does this approach seem feasible with Serverpod? Are there any specific considerations or recommendations you would suggest for implementing this solution?

Thank you,

**Comment by vlidholt** [comment, marked as answer]
Hi!

Serverpod already has an interface for streaming data, you can read about it here:
https://docs.serverpod.dev/concepts/streams

In addition, we are working on greatly improving the streaming interface. This feature will be release with Serverpod 1.2 over the next few weeks. You can read more here:
https://github.com/serverpod/serverpod/issues/2338

I think we already got you covered for your real-time communication needs.
